File: HW/HW12/MinCost.py
from typing import List

class MinCost:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        parent = list(range(n + 1))
        rank = [0] * (n + 1)
        
        def find(u):
            if u != parent[u]:
                parent[u] = find(parent[u])
            return parent[u]

        def union(u, v):
            rootU = find(u)
            rootV = find(v)
            if rootU != rootV:
                if rank[rootU] > rank[rootV]:
                    parent[rootV] = rootU
                elif rank[rootU] < rank[rootV]:
                    parent[rootU] = rootV
                else:
                    parent[rootV] = rootU
                    rank[rootU] += 1
                return True
            return False

        # Start with the cost of wells and add them as edges from a virtual node 0
        edges = [(cost, 0, i + 1) for i, cost in enumerate(wells)]
        # Add the pipes as edges between houses
        edges.extend((cost, house1, house2) for house1, house2, cost in pipes)
        # Sort the edges by cost
        edges.sort()

        total_cost = 0
        # Go through the sorted edges and join the sets
        for cost, u, v in edges:
            if union(u, v):
                total_cost += cost

        return total_cost

File: HW/HW12/MinCost.java
package HW12;
import java.util.Arrays;

public class MinCost {

    private int[] parent;
    private int[] rank;

    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        parent = new int[n + 1];
        rank = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }

        int[][] edges = new int[pipes.length + n][];

        for (int i = 0; i < wells.length; i++) {
            edges[i] = new int[]{0, i + 1, wells[i]};
        }

        for (int i = 0; i < pipes.length; i++) {
            edges[i + n] = new int[]{pipes[i][0], pipes[i][1], pipes[i][2]};
        }

        Arrays.sort(edges, (a, b) -> a[2] - b[2]);

        int totalCost = 0;
        for (int[] edge : edges) {
            if (union(edge[0], edge[1])) {
                totalCost += edge[2];
            }
        }

        return totalCost;
    }

    private int find(int x) {
        if (x != parent[x]) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootX;
            } else if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
            return true;
        }
        return false;
    }
}
